src/mongocxx/options/private/apm.hh

```diff
@@ -37,7 +37,7 @@ static void exception_guard(const char* source, Fn fn) noexcept {
     } catch (...) {
         std::cerr << "fatal error: APM callback " << source << " exited via an exception"
                   << std::endl;
-        std::terminate();
+        throw;
```

Comment:
Revert to using `std::terminate()`? The `throw` results in a warning causing Evergreen task failures:

```
[2023/08/31 20:59:06.981] /data/mci/a62ad8c4c1a22cdebf42866d38f8d790/mongo-cxx-driver/src/mongocxx/../mongocxx/options/private/apm.hh:40:9: error: 'throw' will always call 'terminate' [-Werror=terminate]
[2023/08/31 20:59:06.981]    40 |         throw;
```

Reply:
The intent was to use `noexcept` to invoke `std::terminate()` to take advantage of implicit code paths that might not apply to an explicit call to `std::terminate()`, but I guess that wasn't sufficiently motivated.